staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -165,6 +212,121 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+	_, regionBeta = pv.Labels[v1.LabelFailureDomainBetaRegion]
+	if zoneBeta || regionBeta {
+		// Beta label exists, NodeAffinity not exist, GA label not exists
+		return false, v1.LabelFailureDomainBetaZone, v1.LabelFailureDomainBetaRegion
+	}
+
+	// No labels or NodeAffinity exist, default to GA version
+	return true, v1.LabelTopologyZone, v1.LabelTopologyRegion
+}
+
+// TopologyKeyExist checks if a certain key exists in a VolumeNodeAffinity
+func TopologyKeyExist(key string, vna *v1.VolumeNodeAffinity) bool {
+	if vna == nil || vna.Required == nil || vna.Required.NodeSelectorTerms == nil || len(vna.Required.NodeSelectorTerms) == 0 {
+		return false
+	}
+
+	for _, nodeSelectorTerms := range vna.Required.NodeSelectorTerms {
+		nsrequirements := nodeSelectorTerms.MatchExpressions
+		for _, nodeSelectorRequirement := range nsrequirements {
+			if nodeSelectorRequirement.Key == key {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// translateTopologyFromCSIToInTree
+// 1. Remove any existing CSI topologyKey from NodeAffinity
+// 2. Add Kubernetes Topology in the NodeAffinity if it does not exist
+// 3. Add Kubernetes Topology labels
+func translateTopologyFromCSIToInTree(pv *v1.PersistentVolume, topologyKey string) error {
+
+	csiTopologyZoneValues := getTopologyValues(pv, topologyKey)
+	csiTopologyZoneValStr := strings.Join(csiTopologyZoneValues, labelMultiZoneDelimiter)
+
+	// 1. Frist remove the CSI topology Key
+	removeTopology(pv, topologyKey)
+
+	_, zoneLabel, regionLabel := getTopologyLabel(pv)
+	zoneLabelVal, zoneOK := pv.Labels[zoneLabel]
+	regionLabelVal, regionOK := pv.Labels[regionLabel]
+
+	// 2. Add Kubernetes Topology in the NodeAffinity if it does not exist
+	// Check if Kubernetes Topology already exist
+	topologyZoneValue := getTopologyValues(pv, zoneLabel)
+	if len(topologyZoneValue) == 0 {
+		// No Kubernetes Topology exist in the current PV, we need to add it
+		if zoneOK {
+			zones := strings.Split(zoneLabelVal, labelMultiZoneDelimiter)
+			if len(zones) > 0 {
+				err := addTopology(pv, zoneLabel, zones)
+				if err != nil {
+					return fmt.Errorf("Failed to add Kubernetes topology zone to PV NodeAffinity. %v", err)
+				}
+			}
+		} else {
+			// nothing is in the label, we try our best to add the CSI topology zone in
+			if len(csiTopologyZoneValues) > 0 {
+				err := addTopology(pv, zoneLabel, csiTopologyZoneValues)
+				if err != nil {
+					return fmt.Errorf("Failed to add Kubernetes topology zone to PV NodeAffinity. %v", err)
+				}
+			}
+		}
+		if regionOK {
+			err := addTopology(pv, regionLabel, []string{regionLabelVal})
+			if err != nil {
+				return fmt.Errorf("Failed to add Kubernetes topology region to PV NodeAffinity. %v", err)
+			}
+		}
+	}
+
+	// 3. Add Kubernetes Topology labels, if it already exists, we trust it
+	if !zoneOK {
+		pv.Labels[zoneLabel] = csiTopologyZoneValStr
+	}
+	// Unless we are sure we can derive the region info from zone info. Let's leave the region info for now.
```

Comment:
Since region is more complicated to handle, what if we rename this function to handle only zone, and leave it up to each plugin to handle region as they wish?

Reply:
oops... sorry I did not see this, I have already adopt the function in a way to accept a regionParser function to handle region. And if nil is passing, it will skip the region part. Please let me know if that makes sense or not.